lib/charms/data_platform_libs/v0/object_storage.py

```diff
@@ -0,0 +1,262 @@
+        """Return the azure storage connection info as a dictionary."""
+        for relation in self.relations:
+            if relation and relation.app:
+                return self.relation_data.fetch_relation_data([relation.id])[relation.id]
+        return {}
+
+    def _on_relation_changed_event(self, event: RelationChangedEvent) -> None:
+        """Notify the charm about the presence of Azure credentials."""
+        logger.info(f"Azure storage relation ({event.relation.name}) changed...")
+
+        diff = self._diff(event)
+        if any(newval for newval in diff.added if self.relation_data._is_secret_field(newval)):
+            self.relation_data._register_secrets_to_relation(event.relation, diff.added)
+
+        # check if the mandatory options are in the relation data
+        contains_required_options = True
+        credentials = self.get_azure_connection_info()
+        missing_options = []
+        for configuration_option in AZURE_STORAGE_REQUIRED_INFO:
+            if configuration_option not in credentials:
+                contains_required_options = False
+                missing_options.append(configuration_option)
+
+        # emit credential change event only if all mandatory fields are present
+        if contains_required_options:
+            getattr(self.on, "credentials_changed").emit(
+                event.relation, app=event.app, unit=event.unit
+            )
+        else:
+            logger.warning(
+                f"Some mandatory fields: {missing_options} are not present, do not emit credential change event!"
+            )
+
+    def _on_secret_changed_event(self, event: SecretChangedEvent):
+        """Event handler for handling a new value of a secret."""
+        if not event.secret.label:
+            return
+
+        relation = self.relation_data._relation_from_secret_label(event.secret.label)
+        if not relation:
+            logging.info(
+                f"Received secret {event.secret.label} but couldn't parse, seems irrelevant."
+            )
+            return
+
+        if relation.app == self.charm.app:
+            logging.info("Secret changed event ignored for Secret Owner")
+
+        remote_unit = None
+        for unit in relation.units:
+            if unit.app != self.charm.app:
+                remote_unit = unit
+
+        # check if the mandatory options are in the relation data
+        contains_required_options = True
+        credentials = self.get_azure_connection_info()
+        missing_options = []
+        for configuration_option in AZURE_STORAGE_REQUIRED_INFO:
+            if configuration_option not in credentials:
+                contains_required_options = False
+                missing_options.append(configuration_option)
+
+        # emit credential change event only if all mandatory fields are present
+        if contains_required_options:
+            getattr(self.on, "credentials_changed").emit(
```

Comment:
I would suggest to use the `storage_connection_info_changed` event instead (see comment above).
That would allow us to handle all connection info changes seamlessly.


Alternatively it's opening up another box.
In case we would like to distinguish between 'credentials' and 'other connection info data', then the following applies:

This may sound as a small thing but it's more important than what it seems.

In Data Interfaces we already have a corrrespondant for this event. As for now it's only used by the OpenSearch interface however we are planning to extend it to all interfaces (so that passwords could be changed by `data-integrator`, etc.).

That event is called `authentication_changed`.

We should have the same naming convention for the same purpose. I would suggest to call this event `authentication_changed` insead of `credentials_changed`, as it's likely to introduce incompatibilty or further complications in the future otherwise.

Reply:
I like the idea of changing the name of the events to `storage_connection_info_changed` and `storage_connection_info_requested`. As far as distinguishing the secret-key and other connection parameters, I don't think the separation is an absolute necessary because the client would need all parameters together when making the new connection anyway, and secret-key parameter independently has no significance without the accompanying parameters.

Reply:
Renamed the custom events.